File: utilslib/src/main/java/com/ashlikun/utils/other/ClassUtils.java
package com.ashlikun.utils.other;

import java.lang.annotation.Annotation;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.LinkedList;

/**
 * 作者　　: 李坤
 * 创建时间: 2017/8/7　14:17
 * 邮箱　　：496546144@qq.com
 * <p>
 * 功能介绍：Class反射工具
 */

public class ClassUtils {
    /**
     * 获取全部Field，包括父类
     *
     * @param claxx
     * @return
     */
    public static LinkedList<Field> getAllDeclaredFields(Class<?> claxx) {
        // find all field.
        LinkedList<Field> fieldList = new LinkedList<Field>();
        while (claxx != null && claxx != Object.class) {
            Field[] fs = claxx.getDeclaredFields();
            for (int i = 0; i < fs.length; i++) {
                Field f = fs[i];
                if (!isInvalid(f)) {
                    fieldList.addLast(f);
                }
            }
            claxx = claxx.getSuperclass();
        }
        return fieldList;
    }

    /**
     * 获取域的泛型类型，如果不带泛型返回null
     */
    public static Class<?> getGenericType(Type type) {
        if (type instanceof ParameterizedType) {
            type = ((ParameterizedType) type).getActualTypeArguments()[0];
            if (type instanceof Class<?>) {
                return (Class<?>) type;
            }
        } else if (type instanceof Class<?>) {
            return (Class<?>) type;
        }
        return null;
    }


    /**
     * 是静态常量或者内部结构属性
     */
    public static boolean isInvalid(Field f) {
        return (Modifier.isStatic(f.getModifiers()) && Modifier.isFinal(f.getModifiers())) || f.isSynthetic();
    }

    /**
     * 是否有Class注解
     *
     * @param clazz           a {@link java.lang.Class} object.
     * @param annotationClass a {@link java.lang.Class} object.
     * @return a boolean.
     */
    public static boolean hasClassAnnotation(Class<?> clazz, Class<? extends Annotation> annotationClass) {
        return getClassAnnotation(clazz, annotationClass) != null;
    }

    /**
     * 是否有字段注解
     *
     * @param clazz           a {@link java.lang.Class} object.
     * @param annotationClass a {@link java.lang.Class} object.
     * @param fieldName       a {@link java.lang.String} object.
     * @return a boolean.
     */
    public static boolean hasFieldAnnotation(Class<?> clazz,
                                             Class<? extends Annotation> annotationClass, String fieldName) throws Exception {
        return getFieldAnnotation(clazz, annotationClass, fieldName) != null;
    }

    /**
     * 是否有字段注解
     *
     * @param clazz           a {@link java.lang.Class} object.
     * @param annotationClass a {@link java.lang.Class} object.
     * @param methodName      a {@link java.lang.String} object.
     * @param paramType       a {@link java.lang.Class} object.
     * @return a boolean.
     */
    public static boolean hasMethodAnnotation(Class<?> clazz,
                                              Class<? extends Annotation> annotationClass, String methodName, Class<?>... paramType) throws Exception {
        return getMethodAnnotation(clazz, annotationClass, methodName, paramType) != null;
    }

    /**
     * 获取类注解
     *
     * @param clazz           类
     * @param annotationClass 注解类
     * @return a A object.
     */


    public static <A extends Annotation> A getClassAnnotation(Class<?> clazz, Class<A> annotationClass) {
        return clazz.getAnnotation(annotationClass);

    }

    /**
     * 获取类成员注解
     *
     * @param clazz           类
     * @param annotationClass 注解类
     * @param fieldName       成员属性名
     * @return a A object.
     */


    public static <A extends Annotation> A getFieldAnnotation(Class<?> clazz,
                                                              Class<A> annotationClass, String fieldName) throws Exception {
        try {
            Field field = clazz.getDeclaredField(fieldName);
            if (field == null) {

                throw new Exception("no such field[" + fieldName + "] in " + clazz.getCanonicalName());

            }
            return field.getAnnotation(annotationClass);

        } catch (SecurityException e) {
            e.printStackTrace();

            throw new Exception("access error: field[" + fieldName + "] in " + clazz.getCanonicalName(), e);

        } catch (NoSuchFieldException e) {
            e.printStackTrace();

            throw new Exception("no such field[" + fieldName + "] in " + clazz.getCanonicalName());

        }

    }

    /**
     * 获取类方法上的注解
     *
     * @param clazz           类
     * @param annotationClass 注解类
     * @param methodName      方法名
     * @param paramType       方法参数
     * @return a A object.
     */
    public static <A extends Annotation> A getMethodAnnotation(Class<?> clazz,
                                                               Class<A> annotationClass, String methodName, Class<?>... paramType)
            throws Exception {
        try {
            Method method = clazz.getDeclaredMethod(methodName, paramType);
            if (method == null) {
                throw new Exception("access error: method[" + methodName + "] in " + clazz.getCanonicalName());
            }
            return method.getAnnotation(annotationClass);
        } catch (SecurityException e) {
            e.printStackTrace();
            throw new Exception("access error: method[" + methodName + "] in " + clazz.getCanonicalName(), e);
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
            throw new Exception("no such method[" + methodName + "] in " + clazz.getCanonicalName(), e);
        }
    }
}

File: utilslib/src/main/java/com/ashlikun/utils/other/ClassUtils.kt
package com.ashlikun.utils.other

import java.lang.reflect.*

/**
 * @author　　: 李坤
 * 创建时间: 2021/12/12 16:53
 * 邮箱　　：496546144@qq.com
 *
 *
 * 功能介绍：Class反射工具
 */
/**
 * 获取全部Field，包括父类
 */
inline fun Class<*>.getAllFields() = ClassUtils.getAllFields(this)

/**
 * 获取全部Method，包括父类
 */
inline fun Class<*>.getAllMethods() = ClassUtils.getAllMethods(this)

/**
 * 获取指定的字段
 */
inline fun Class<*>.getField(fieldName: String) = ClassUtils.getField(this, fieldName)
inline fun Class<*>.getMethod(fieldName: String, vararg parameterTypes: Class<*>?) =
    ClassUtils.getMethod(this, fieldName, *parameterTypes)

/**
 * 反射字段 设置值
 */
inline fun Any.getFieldValue(fieldName: String) = ClassUtils.getFieldValue(this, fieldName)
inline fun Any.setFieldValue(fieldName: String, value: Any?) =
    ClassUtils.setFieldValue(this, fieldName, value)

/**
 * 反射方法
 *
 * @param object      要反射的对象
 * @param methodNames 要反射的方法名称
 * @param parameterTypes 参数类型
 * @param args 值
 */
inline fun Any.callMethod(
    methodName: String, parameterTypes: Array<Class<*>?>? = null,
    vararg args: Any?
) = ClassUtils.callMethod(this, methodName, parameterTypes, *args)

inline fun Class<*>.callStaticMethod(
    methodName: String, parameterTypes: Array<Class<*>?>? = null,
    vararg args: Any?
) = ClassUtils.callStaticMethod(this, methodName, parameterTypes, *args)

/**
 * 获取泛型类型 第一个，如果不带泛型返回null,
 */
inline fun Class<*>.getGenericType() = ClassUtils.getGenericType(this)

object ClassUtils {
    /**
     * 获取全部Field，包括父类
     */
    fun getAllFields(claxx: Class<*>): MutableList<Field> {
        var claxx: Class<*>? = claxx
        val fieldList = mutableListOf<Field>()
        while (claxx != null && claxx != Any::class.java) {
            val fs = claxx.declaredFields
            for (i in fs.indices) {
                val f = fs[i]
                if (!isInvalid(f)) {
                    fieldList.add(f)
                }
            }
            claxx = claxx.superclass
        }
        return fieldList
    }

    /**
     * 获取指定的字段
     */
    fun getField(claxx: Class<*>, fieldName: String): Field? {
        var claxx: Class<*>? = claxx
        if (fieldName.isEmpty()) {
            return null
        }
        while (claxx != null && claxx != Any::class.java) {
            try {
                val f = claxx.getDeclaredField(fieldName)
                if (f != null) {
                    return f
                }
            } catch (e: NoSuchFieldException) {
                e.printStackTrace()
            }
            claxx = claxx.superclass
        }
        return null
    }

    /**
     *  反射字段 的值
     *
     * @param object    要反射的对象
     * @param fieldName 要反射的字段名称
     */
    fun getFieldValue(obj: Any, fieldName: String): Any? {
        if (fieldName.isEmpty()) {
            return null
        }
        try {
            val field = getField(obj.javaClass, fieldName)
            if (field != null) {
                field.isAccessible = true
                return field[obj]
            }
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
        return null
    }

    /**
     * 反射字段 设置值
     *
     * @param object    要反射的对象
     * @param fieldName 要反射的字段名称
     */
    fun setFieldValue(obj: Any, fieldName: String, value: Any?): Field? {
        if (fieldName.isEmpty()) {
            return null
        }
        try {
            val field = getField(obj.javaClass, fieldName)
            if (field != null) {
                field.isAccessible = true
                field[obj] = value
                return field
            }
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
        return null
    }

    /**
     * 获取全部Method，包括父类
     */
    fun getAllMethods(claxx: Class<*>): MutableList<Method> {
        var claxx: Class<*>? = claxx
        val fieldList = mutableListOf<Method>()
        while (claxx != null && claxx != Any::class.java) {
            val fs = claxx.declaredMethods
            for (i in fs.indices) {
                val f = fs[i]
                fieldList.add(f)
            }
            claxx = claxx.superclass
        }
        return fieldList
    }

    /**
     * 获取指定的方法
     */
    fun getMethod(
        claxx: Class<*>, methodName: String, vararg parameterTypes: Class<*>?
    ): Method? {
        var claxx: Class<*>? = claxx
        if (methodName.isEmpty()) {
            return null
        }
        while (claxx != null && claxx != Any::class.java) {
            try {
                val f = claxx.getDeclaredMethod(methodName, *parameterTypes)
                if (f != null) {
                    return f
                }
            } catch (e: NoSuchMethodException) {
                e.printStackTrace()
            }
            claxx = claxx.superclass
        }
        return null
    }

    /**
     * 反射方法
     *
     * @param object      要反射的对象
     * @param methodNames 要反射的方法名称
     * @param parameterTypes 参数类型
     * @param args 值
     */
    fun callMethod(
        obj: Any,
        methodName: String,
        parameterTypes: Array<Class<*>?>? = null,
        vararg args: Any?
    ): Any? {
        if (methodName.isEmpty()) {
            return null
        }
        try {
            val method = if (parameterTypes == null) getMethod(
                obj.javaClass,
                methodName
            ) else getMethod(obj.javaClass, methodName, *parameterTypes)
            if (method != null) {
                method.isAccessible = true
                return method.invoke(obj, *args)
            }
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
        return null
    }


    /**
     * 反射静态方法
     *
     * @param claxx          要反射的对象
     * @param methodName     要反射的方法名称
     * @param parameterTypes 参数类型
     * @param args           参数
     * @return
     */
    fun callStaticMethod(
        claxx: Class<*>,
        methodName: String,
        parameterTypes: Array<Class<*>?>?,
        vararg args: Any?
    ): Any? {
        if (methodName.isEmpty()) {
            return null
        }
        try {
            val method = getMethod(claxx, methodName, *parameterTypes!!)
            if (method != null) {
                method.isAccessible = true
                return method.invoke(null, *args)
            }
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
        return null
    }

    /**
     * 获取泛型类型 第一个，如果不带泛型返回null,
     */
    fun getGenericType(cls: Class<*>): Type? {
        val types: Type? = cls.genericSuperclass
        var parentypes: Array<Type?> //泛型类型集合
        if (types is ParameterizedType) {
            parentypes = types.actualTypeArguments
        } else {
            parentypes = cls.getGenericInterfaces()
            for (childtype in parentypes) {
                if (childtype is ParameterizedType) {
                    val rawType = childtype.rawType
                    //实现的接口是Callback
                    if (rawType is Class<*>) {
                        //Callback里面的类型
                        parentypes = childtype.actualTypeArguments
                    }
                }
            }
        }
        return parentypes.getOrNull(0)
    }

    /**
     * 是静态常量或者内部结构属性
     */
    fun isInvalid(f: Field) =
        Modifier.isStatic(f.modifiers) && Modifier.isFinal(f.modifiers) || f.isSynthetic

    /**
     * 是否有Class注解
     */
    fun hasClassAnnotation(clazz: Class<*>, annotationClass: Class<out Annotation>) =
        getClassAnnotation(clazz, annotationClass) != null

    /**
     * 是否有字段注解
     */
    fun hasFieldAnnotation(
        clazz: Class<*>,
        annotationClass: Class<out Annotation>?, fieldName: String
    ) = getFieldAnnotation(clazz, annotationClass, fieldName) != null

    /**
     * 是否有字段注解
     *
     * @param clazz           a [Class] object.
     * @param annotationClass a [Class] object.
     * @param methodName      a [String] object.
     * @param paramType       a [Class] object.
     * @return a boolean.
     */
    fun hasMethodAnnotation(
        clazz: Class<*>,
        annotationClass: Class<out Annotation>?, methodName: String, vararg paramType: Class<*>?
    ) = getMethodAnnotation(clazz, annotationClass, methodName, *paramType) != null

    /**
     * 获取类注解
     *
     * @param clazz           类
     * @param annotationClass 注解类
     * @return a A object.
     */
    fun <A : Annotation?> getClassAnnotation(clazz: Class<*>, annotationClass: Class<A>): A? {
        return clazz.getAnnotation(annotationClass)
    }

    /**
     * 获取类成员注解
     *
     * @param clazz           类
     * @param annotationClass 注解类
     * @param fieldName       成员属性名
     * @return a A object.
     */
    fun <A : Annotation> getFieldAnnotation(
        clazz: Class<*>,
        annotationClass: Class<A>?, fieldName: String
    ): A? {
        return try {
            val field = clazz.getDeclaredField(fieldName)
            field.getAnnotation(annotationClass)
        } catch (e: SecurityException) {
            e.printStackTrace()
            null
        } catch (e: NoSuchFieldException) {
            e.printStackTrace()
            null
        }
    }

    /**
     * 获取类方法上的注解
     *
     * @param clazz           类
     * @param annotationClass 注解类
     * @param methodName      方法名
     * @param paramType       方法参数
     * @return a A object.
     */
    fun <A : Annotation> getMethodAnnotation(
        clazz: Class<*>,
        annotationClass: Class<A>?, methodName: String, vararg paramType: Class<*>?
    ): A? {
        return try {
            val method = clazz.getDeclaredMethod(methodName, *paramType)
            method.getAnnotation(annotationClass)
        } catch (e: SecurityException) {
            e.printStackTrace()
            null
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
            null
        }
    }
}